python/mobitex_deframer.py

```diff
@@ -0,0 +1,366 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+# Copyright 2025 Fabian P. Schmidt <kerel@mailbox.org>
+#
+# This file is part of gr-satellites
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+#
+
+import itertools
+
+from gnuradio import gr
+import pmt
+import numpy as np
+
+from .check_eseo_crc import crc16_ccitt_zero as crc16_ccitt_zero
+from .mobitex_fec import decode, Status
+
+
+def decode_control_bytes(control_bytes: list[int], control_byte_crc: int):
+    # Error Correction of the control bytes
+    control_bytes[0], fec0, status0 = decode(
+        (control_bytes[0] << 4) | (control_byte_crc >> 4)
+    )
+    control_bytes[1], fec1, status1 = decode(
+        (control_bytes[1] << 4) | (control_byte_crc & 0x0F)
+    )
+    control_byte_crc = fec0 << 4 | fec1
+
+    # Decode control bytes
+    result = {
+        # 1st byte, bits 0-4 (+ 1)
+        "num_data_blocks": (control_bytes[0] & 0x1F) + 1,
+        # 1st byte, bits 5-7
+        "message_type": (control_bytes[0] >> 5) & 0x07,
+        # 2nd byte, bit 0
+        "baud_bit": control_bytes[1] & 0x01,
+        # 2nd byte,bit 1
+        "ack_bit": (control_bytes[1] >> 1) & 0x01,
+        # 2nd byte,bits 2-3
+        "sub_address": (control_bytes[1] >> 2) & 0x03,
+        # 2nd byte,bits 4-7
+        "address": (control_bytes[1] >> 4) & 0x0F,
+    }
+
+    if (
+        status0 == Status.ERROR_UNCORRECTABLE or
+        status1 == Status.ERROR_UNCORRECTABLE
+    ):
+        raise ValueError("Control bytes FEC failed.")
+
+    # Count bit errors
+    bit_errors = 0
+    if status0 == Status.ERROR_CORRECTED:
+        bit_errors += 1
+    if status1 == Status.ERROR_CORRECTED:
+        bit_errors += 1
+
+    return result, bit_errors, control_bytes, control_byte_crc
+
+
+def decode_unknown_callsign(
+    callsign_bytes: bytes,
+    crc_bytes: bytes,
+    max_bit_flips: int,
+    callsign_ref: bytes = None,
+) -> tuple[bytes, int]:
+    """
+    Attempt to error-correct a callsign by flipping bits until CRC matches.
+
+    Args:
+        callsign_bytes: The callsign bytes to decode
+        crc_bytes: The CRC bytes to verify against
+        max_bit_flips: Maximum number of bits to flip during error correction
+        callsign_ref: Unused argument
+
+    Returns:
+        tuple containing:
+            - corrected callsign bytes
+            - number of bit errors corrected
+
+    Raises:
+        ValueError: If no valid callsign could be found
+                    within max_bit_flips bit flips
+    """
+    if max_bit_flips <= 0:
+        # Do not attempt error correction
+        computed_crc = crc16_ccitt_zero(bytes(callsign_bytes))
+        computed_crc_bytes = computed_crc.to_bytes(length=2)
```

Comment:
According to [the docs](https://docs.python.org/3/library/stdtypes.html#int.to_bytes), the default value for the `byteorder` argument of `to_bytes()` was added in Python 3.11, so it looks like this will need an explicit `byteorder='big'` to work with Python < 3.11.